cognito/cognito.go

```diff
@@ -0,0 +1,312 @@
+
+	input := &cognitoidentityprovider.AdminResetUserPasswordInput{
+		UserPoolId: &ca.poolID,
+		Username:   &username,
+	}
+
+	request, _ := ca.provider.AdminResetUserPasswordRequest(input)
+	if err := request.Send(); err != nil {
+		return wrapOpsErr(err, "error in cognito.Adapter while sending AdminResetUserPasswordRequest")
+	}
+	return nil
+}
+
+func (ca *Adapter) ConfirmForgotPassword(username, newPassword, confirmationCode string) error {
+
+	secretHash, err := ca.generateSecretHash(username)
+	if err != nil {
+		return wrapErr(err, ErrSecretHashEncoding, "Cannot encode secret hash")
+	}
+
+	input := &cognitoidentityprovider.ConfirmForgotPasswordInput{
+		ClientId:         &ca.clientID,
+		ConfirmationCode: &confirmationCode,
+		Password:         &newPassword,
+		SecretHash:       &secretHash,
+		Username:         &username,
+	}
+
+	request, _ := ca.provider.ConfirmForgotPasswordRequest(input)
+	if err := request.Send(); err != nil {
+		return wrapOpsErr(err, "error in cognito.Adapter while sending ConfirmForgotPasswordRequest")
+	}
+	return nil
+}
+
+func (ca *Adapter) respondToAuthChallenge(username, password string,
+	session *string) (*cognitoidentityprovider.AdminRespondToAuthChallengeOutput, error) {
+
+	secretHash, err := ca.generateSecretHash(username)
+	if err != nil {
+		return nil, wrapErr(err, ErrSecretHashEncoding, "Cannot encode secret hash")
+	}
+
+	challengeName := cognitoidentityprovider.ChallengeNameTypeNewPasswordRequired
+
+	input := &cognitoidentityprovider.AdminRespondToAuthChallengeInput{
+		ChallengeName: &challengeName,
+		ChallengeResponses: map[string]*string{
+			"USERNAME":     &username,
+			"NEW_PASSWORD": &password,
+			"SECRET_HASH":  &secretHash,
+		},
+		ClientId:   &ca.clientID,
+		UserPoolId: &ca.poolID,
+		Session:    session,
+	}
+
+	request, output := ca.provider.AdminRespondToAuthChallengeRequest(input)
+	if err := request.Send(); err != nil {
+		return nil, wrapOpsErr(err, "error in cognito.Adapter while sending AdminRespondToAuthChallengeRequest")
+	}
+	return output, nil
+}
+
+func (ca *Adapter) changePassword(oldPassword, newPassword string, token *string) (*cognitoidentityprovider.ChangePasswordOutput, error) {
+
+	input := &cognitoidentityprovider.ChangePasswordInput{
+		AccessToken:      token,
+		PreviousPassword: &oldPassword,
+		ProposedPassword: &newPassword,
+	}
+
+	request, output := ca.provider.ChangePasswordRequest(input)
+	if err := request.Send(); err != nil {
+		return nil, wrapOpsErr(err, "error in cognito.Adapter while sending ChangePasswordRequest")
+	}
+	return output, nil
+}
+
+func (ca *Adapter) generateSecretHash(username string) (string, error) {
+	mac := hmac.New(sha256.New, []byte(ca.clientSecret))
+	_, err := mac.Write([]byte(username + ca.clientID))
+	if err != nil {
+		return "", err
+	}
+	return base64.StdEncoding.EncodeToString(mac.Sum(nil)), nil
+}
+
+func (ca *Adapter) fromAttributes(attrs []*cognitoidentityprovider.AttributeType) map[string]string {
```

Comment:
Is it necessary to define method on adapter? maybe func would be enough? What do you think?

Reply:
true